dynawo/sources/Models/CPP/Controls/Shunts/ModelCentralizedShuntsSectionControl/DYNModelCentralizedShuntsSectionControl.cpp

```diff
@@ -0,0 +1,346 @@
+      tNext_ = findParameterDynamic("tNext").getValue<double>();
+      for (int s = 0; s < nbShunts_; ++s) {
+        std::stringstream section0Name;
+        section0Name << "section0_" << s;
+        sections0_.push_back(findParameterDynamic(section0Name.str()).getValue<int>());
+        std::stringstream SectionMinName;
+        SectionMinName << "SectionMin_" << s;
+        SectionsMin_.push_back(findParameterDynamic(SectionMinName.str()).getValue<int>());
+        std::stringstream SectionMaxName;
+        SectionMaxName << "SectionMax_" << s;
+        SectionsMax_.push_back(findParameterDynamic(SectionMaxName.str()).getValue<int>());
+        std::stringstream DeadbandUPuName;
+        DeadbandUPuName << "DeadbandUPu_" << s;
+        DeadbandsUPu_.push_back(findParameterDynamic(DeadbandUPuName.str()).getValue<double>());
+      }
+    } catch (const DYN::Error& e) {
+    Trace::error() << e.what() << Trace::endline;
+    throw DYNError(Error::MODELER, NetworkParameterNotFoundFor, staticId());
+    }
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::defineVariables(std::vector<boost::shared_ptr<Variable> >& variables) {
+    // only one UMonitoredPu variable shared among all shunts as all of them are regulating the same node
+    variables.push_back(VariableNativeFactory::createState("UMonitoredPu_value", CONTINUOUS));
+    for (int s = 0; s < nbShunts_; ++s) {
+      std::stringstream URefPuName;
+      URefPuName << "URefPu_" << s << "_value";
+      variables.push_back(VariableNativeFactory::createState(URefPuName.str(), DISCRETE));
+    }
+    for (int s = 0; s < nbShunts_; ++s) {
+      std::stringstream sectionName;
+      sectionName << "section_" << s << "_value";
+      variables.push_back(VariableNativeFactory::createState(sectionName.str(), DISCRETE));
+    }
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::init(const double /*t0*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::collectSilentZ(BitMask* /*silentZTable*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::getSize() {
+    sizeF_ = 0;
+    sizeY_ = 1;  // UMonitoredPu
+    sizeZ_ = 2 * nbShunts_;  // URefPu values are stored at indexes [0, nbShunts[, sections are stored at [nbShunts, 2 * nbShunts[
+    sizeG_ = 4;
+    sizeMode_ = 0;
+    calculatedVars_.assign(nbCalculatedVariables_, 0);
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalStaticYType() {
+    yType_[0] = EXTERNAL;
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalF(double /*t*/, propertyF_t /*type*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalJt(const double /*t*/, const double /*cj*/, SparseMatrix& /*jt*/, const int /*rowOffset*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalG(const double t) {
+    if (!isSelf_) {
+      for (int s = 0; s < nbShunts_; ++s) {
+        if (doubleNotEquals(yLocal_[0], (zLocal_[s] - DeadbandsUPu_[s])) &&
```

Comment:
create double UMonitoredPu = yLocal_[0]; outside of for loop. create double minValue = zLocal_[s] - DeadbandsUPu_[s]; and use it in various conditons. same for zLocal_[s] + DeadbandsUPu_[s]